centreon/lib/perl/centreon/common/centreonvault.pm

```diff
@@ -0,0 +1,412 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+package centreon::common::centreonvault;
+
+use strict;
+use warnings;
+
+use MIME::Base64;
+use Crypt::OpenSSL::AES;
+use Net::Curl::Easy qw(:constants);
+use JSON::XS;
+
+my $VAULT_PATH_REGEX = qr/^secret::hashicorp_vault::([^:]+)::(.+)$/;
+
+sub new {
+    my ($class, %options) = @_;
+    my $self  = bless \%options, $class;
+    # mandatory options:
+    # - logger: logger object
+    # - config_file: either path of a JSON vault config file or the configuration as a perl hash.
+
+    $self->{enabled} = 1;
+    $self->{crypted_credentials} = 1;
+
+    if ( !$self->init() ) {
+        $self->{enabled} = 0;
+        $self->{logger}->writeLogError("An error occurred in init() method. Centreonvault cannot be used.");
+    }
+    return $self;
+}
+
+
+sub init {
+    my ($self, %options) = @_;
+
+    $self->check_options() or return undef;
+
+    # for unit test purpose, if the config is given as an hash, we don't try to read the config file.
+    if (ref $self->{config_file} eq ref {}) {
+        $self->{vault_config} = $self->{config_file};
+    } else {
+        # check if the following information is available
+        $self->{logger}->writeLogDebug("Reading Vault configuration from file " . $self->{config_file} . ".");
+        $self->{vault_config} = parse_json_file('json_file' => $self->{config_file});
+        if (defined($self->{vault_config}->{error_message})) {
+            $self->{logger}->writeLogError("Error while parsing " . $self->{config_file} . ": "
+                . $self->{vault_config}->{error_message});
+            return undef;
+        }
+    }
+    $self->check_configuration() or return undef;
+
+    $self->{logger}->writeLogDebug("Vault configuration read. Name: " . $self->{vault_config}->{name}
+        . ". Url: " . $self->{vault_config}->{url} . ".");
+
+    # Create the Curl object, it will be used several times
+    $self->{curl_easy} = Net::Curl::Easy->new();
+    $self->{curl_easy}->setopt( CURLOPT_USERAGENT, "Centreon VMware daemon's centreonvault.pm");
+
+    return 1;
+}
+
+sub check_options {
+    my ($self, %options) = @_;
+
+    if ( !defined($self->{logger}) ) {
+        die "FATAL: No logger given to the constructor. Centreonvault cannot be used.";
+    }
+    if ( !defined($self->{config_file})) {
+        $self->{logger}->writeLogError("No config file given to the constructor. Centreonvault cannot be used.");
+        return undef;
+    }
+    if ( ! -f $self->{config_file} and ref $self->{config_file} ne ref {}) {
+        $self->{logger}->writeLogError("The given configuration file " . $self->{config_file}
+            . " does not exist. Centreonvault cannot be used.");
```

Comment:
After some user feedback, I made the error message [less alarming](https://github.com/centreon/centreon-plugins/blob/39d6d4bbf96c9b99c229896e9b98e8c19b2704ed/connectors/vmware/src/centreon/script/centreonvault.pm#L89). If you feel like doing so...